### PR TITLE
consolidate node deletion logic between kube-controller-manager and cloud-controller-manager

### PR DESCRIPTION
**Pull request by andrewsykim**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Consolidates node deletion logic between kube-controller-manager and cloud-controller-manager by creating a new "cloud node lifecycle" controller. This new controller pulls the node deletion logic from both the node lifecycle controller (KCM) and the cloud node controller (CCM) which more or less have the same logic with a _few_ exceptions. The exceptions are mostly differences in semantics (different method names, errors, using informers instead of client, etc) and not actual logic. 

This PR is needed for several reasons:
* avoid divergence of node deletion logic between kube-controller-manager (in-tree) and cloud-controller-manager (out-of-tree).
* removes dependency to cloud providers in node lifecycle controller
* allows us to disable node deletion controller only without deleting node lifecycle controller which is something we need as part of the in-tree -> out-of-tree migration. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/sig cloud-provider
/assign @cheftako

### CODE REVIEW

File: pkg/controller/cloud/node_gc_controller.go
Comment (andrewsykim) [inline pkg/controller/cloud/node_gc_controller.go]: This code is pulled straight out of KCM/CCM

File: pkg/controller/util/node/controller_utils.go
Comment (andrewsykim) [inline pkg/controller/util/node/controller_utils.go]: There's nothing in k8s/k8s that actually checks for this error so removed it.

File: pkg/controller/util/node/controller_utils.go
Comment (andrewsykim) [inline pkg/controller/util/node/controller_utils.go]: replaced by [ensureNodeExistsByProviderID](https://github.com/kubernetes/kubernetes/blob/521028ed52526abd832f8a6c9c5ec62340fccbb2/pkg/controller/cloud/node_controller.go#L446-L467) which is functionally the same but uses providerID if it exists, otherwise uses `InstanceID`.

File: pkg/controller/cloud/node_gc_controller_test.go
Comment (cheftako) [inline pkg/controller/cloud/node_gc_controller_test.go]: Sadly a lot of the cloud providers have not implemented this call yet and so rather than returning true/false they return a not implemented error. Can we have a test for that case? (For GCE for instance it would return error for shutdown but would correctly return true/false for existsByProviderID)

File: pkg/controller/cloud/node_gc_controller.go
Comment (cheftako) [inline pkg/controller/cloud/node_gc_controller.go]: Wouldn't it make more sense to check this in start/run and just skip the controller entirely in that case?

Comment (andrewsykim) [reply pkg/controller/cloud/node_gc_controller.go]: I had the same thought and added it here https://github.com/kubernetes/kubernetes/pull/70344/files#diff-da5354d582a4b27381b8cee161e46163R82, but then figured it can't hurt to check it here too

Comment (andrewsykim) [reply pkg/controller/cloud/node_gc_controller.go]: I did what service controller does, which is to log the error but still return nil so kube-controller-manager doesn't crash. See [here](https://github.com/kubernetes/kubernetes/pull/70344/files#diff-0ca8b7870d367d5fc309a2e737a6eed2R159)
```go
func startCloudNodeGarbageCollectorController(ctx ControllerContext) (http.Handler, bool, error) {
	cloudNodeGCController, err := cloudcontroller.NewCloudNodeGarbageCollectorController(
		ctx.ClientBuilder.ClientOrDie("cloud-node-gc-controller"),
		ctx.Cloud,
		ctx.ComponentConfig.KubeCloudShared.NodeMonitorPeriod.Duration,
	)
	if err != nil {
		// the controller manager should continue to run if the "Instances" interface is not
		// supported, though it's unlikely for a cloud provider to not support it
		glog.Errorf("failed to start cloud node garbage collector controller: %v", err)
		return nil, false, nil
	}
	go cloudNodeGCController.Run(ctx.Stop)
	return nil, true, nil
}
```

File: pkg/controller/cloud/node_gc_controller_test.go
Comment (andrewsykim) [reply pkg/controller/cloud/node_gc_controller_test.go]: done

File: pkg/controller/cloud/node_gc_controller.go
Comment (cheftako) [inline pkg/controller/cloud/node_gc_controller.go]: Minor quibble but I believe the past tense of shutdown is shutdown.
(https://en.wiktionary.org/wiki/shut_down)

File: pkg/controller/cloud/node_gc_controller.go
Comment (cheftako) [reply pkg/controller/cloud/node_gc_controller.go]: Good call

File: pkg/controller/cloud/node_gc_controller.go
Comment (mtaufen) [inline pkg/controller/cloud/node_gc_controller.go]: `nodeutil.RecordNodeEvent` can simplify the event creation/recording ("k8s.io/kubernetes/pkg/controller/util/node")

File: pkg/controller/cloud/node_gc_controller.go
Comment (mtaufen) [inline pkg/controller/cloud/node_gc_controller.go]: It may be more minimal to just directly call `nodeutil.ExistsInCloudProvider(c.cloud, types.NodeName(node.Name))`.

File: pkg/controller/cloud/node_gc_controller.go
Comment (mtaufen) [inline pkg/controller/cloud/node_gc_controller.go]: It may be more minimal to just directly call `nodeutil.ShutdownInCloudProvider(context.TODO(), c.cloud, node)`.

File: pkg/controller/cloud/node_gc_controller.go
Comment (mtaufen) [inline pkg/controller/cloud/node_gc_controller.go]: What if the node was deleted or shut down in the cloud provider before the Kubelet could post a ready status?

It might be more robust to assume `v1.ConditionUnknown` if there is no ready condition on the node yet, and then call directly into the cloud to see if a VM exists/is shut down. Since in the unknown case we'd then always check the cloud provider directly, and only take action when the cloud reports deletion or shut-down, we can probably also avoid the retries/additional Get requests. 

And then if we're also listing from the informer cache, that will help to reduce any additional load on the API server that splitting out these control loops would create.

If we then went a little further and moved the body of this for loop into a helper function, the initial status checks might collapse to something like this, at which point we can un-nest a lot of the subsequent logic:
```
// Default status to v1.ConditionUnknown (in fact, this is the same as the default "fake" status from tryUpdateNodeHealth in the kube node lifecycle controller)
status := v1.ConditionUnknown
if _, c := v1node.GetNodeCondition(&node.Status, v1.NodeReady); c != nil {
  status = c.Status
}

// Skip healthy nodes.
if status == v1.ConditionTrue {
  return
}
```

**Review by mtaufen — CHANGES_REQUESTED** [review]
Thanks for this! We've wanted to do something like this for a long time.

I gave the control loop a first pass, there are still some pieces I have to review and I have yet to look at the tests.

File: pkg/controller/cloud/node_gc_controller.go
Comment (andrewsykim) [reply pkg/controller/cloud/node_gc_controller.go]: That method was removed and copied into this file (but private) because this is the only controller that uses it now.

File: pkg/controller/cloud/node_gc_controller.go
Comment (andrewsykim) [reply pkg/controller/cloud/node_gc_controller.go]: Same here, that method was removed in favour of this method

**Comment by andrewsykim** [comment]
@mtaufen thanks for reviewing! I agree with your comments but my goal here was to introduce the least amount of logical changes while breaking out the node deletion part of each controller. I'm happy to address those comments here but it will be harder to track down bugs later as we will merge everything in the same PR. More than happy to address those comments in a follow up PRs so we can easily revert the logical changes without undoing everything else. Will defer judgement to you though, let me know what you think :)

**Comment by andrewsykim** [comment]
> What if the node was deleted or shut down in the cloud provider before the Kubelet could post a ready status? 

Realizing now that this is likely a bug that we'll want addressed so will address your comments in this PR :)

File: pkg/controller/cloud/node_gc_controller.go
Comment (mtaufen) [reply pkg/controller/cloud/node_gc_controller.go]: Ah, I see, sounds good then

File: pkg/controller/cloud/node_gc_controller.go
Comment (mtaufen) [reply pkg/controller/cloud/node_gc_controller.go]: Ah, I see, sounds good then

File: cmd/cloud-controller-manager/app/controllermanager.go
Comment (mtaufen) [inline cmd/cloud-controller-manager/app/controllermanager.go]: since you mentioned naming I did some brainstorming :)
- cloudHasNodeController
- cloudNodeOnController
- cloudNodeThereController
- cloudNodePresentController
- cloudNodePresenceController
- cloudNodeRosterController
- cloudNodeFinder
- cloudNodeChecker

When I was hacking around with this, I got in the Halloween spirit and called it the NodeReaper... but that's really no better than GarbageCollector...

But GarbageCollector doesn't really fit because this controller can also just taint.

Maybe something like `cloudNodeExistsController` would work? Because the point is to check whether the node is still there, even though the Kubelet isn't reporting? (and the three modes are "there and not working," "there and shut down," or "not there")

File: pkg/controller/cloud/node_controller.go
Comment (mtaufen) [inline pkg/controller/cloud/node_controller.go]: For line 114: Some comments may need to be updated in this file, since now this controller doesn't delete nodes anymore.

**Comment by mtaufen** [comment]
Regarding comments on risk management; how much regression test coverage do we have?

It didn't look to me like we have explicit e2e or integration tests that will exercise at least the following:
```
(Node: Not Ready)      && (Cloud: Node DNE)       -> Something deletes Node
(Node: Not Ready)      && (Cloud: Node Shut Down) -> Something taints Node
(Node: Ready, Tainted) && (Cloud: Running)        -> Something removes taint
(Node: Not Ready)      && (Cloud: Node Exists)    -> Something else happens (depending on other controllers)
```
The advantage of e2e/integration tests would be that we can run them before and after this PR, without changing the tests, to see if we broke anything. Can we add those tests first, then make this change with more confidence?

File: cmd/cloud-controller-manager/app/controllermanager.go
Comment (andrewsykim) [reply cmd/cloud-controller-manager/app/controllermanager.go]: What are your thoughts on having both a `CloudNodeLifecycleController` and a `NodeLifeCycleController`. I like the term "LifeCycle" as I think its accurate as to what the controller is doing but maybe it might be confusing to have controllers with similar names.

**Comment by mtaufen** [comment]
Hey @andrewsykim, how is this coming?

**Comment by andrewsykim** [comment]
@mtaufen sorry for the delay! I would like to get this in for v1.13. I pushed up what I have so far which is all the comments addressed except for adding E2E tests. I'll be partially offline for the next few days traveling for KubeCON Shanghai so please review what I have so far and I'll do my best to address any other concerns before the v1.13 deadline. Thanks!

/milestone v1.13

File: pkg/controller/cloud/node_lifecycle_controller.go
Comment (mtaufen) [inline pkg/controller/cloud/node_lifecycle_controller.go]: Could simplify to `klog.Infof`, I think.

File: pkg/controller/cloud/node_lifecycle_controller.go
Comment (mtaufen) [inline pkg/controller/cloud/node_lifecycle_controller.go]: consider an error instead of a panic?

File: pkg/controller/cloud/node_lifecycle_controller.go
Comment (mtaufen) [inline pkg/controller/cloud/node_lifecycle_controller.go]: s/fale/fail

File: pkg/controller/cloud/node_lifecycle_controller.go
Comment (mtaufen) [inline pkg/controller/cloud/node_lifecycle_controller.go]: If the recorder also logs the event, this is redundant.

File: cmd/cloud-controller-manager/app/controllermanager.go
Comment (mtaufen) [reply cmd/cloud-controller-manager/app/controllermanager.go]: sgtm

File: pkg/controller/cloud/node_lifecycle_controller_test.go
Comment (mtaufen) [inline pkg/controller/cloud/node_lifecycle_controller_test.go]: maybe test with exists in addition to does-not-exist?

File: pkg/controller/cloud/node_lifecycle_controller_test.go
Comment (mtaufen) [inline pkg/controller/cloud/node_lifecycle_controller_test.go]: Please use the field names, these struct literals are fairly large and difficult to read without the names included.

**Comment by mtaufen** [comment]
Nice work.
Looks pretty good overall modulo additional tests discussed above.
Left some comments on minor nits here and there.

**Comment by andrewsykim** [comment]
@mtaufen addressed comments! I started working on E2E tests on a separate branch and will likely open a separate PR for it. I think E2E tests for this case is going to take some time as there are some things that need to be considered:
* Deleting a cloud provider's node has implications for the rest of the E2E tests (cc @BenTheElder). Especially if we are running tests in parallel. 
* Building cloud API authentication, etc into the E2E testing framework. We can't delete nodes via the kubernetes control plane since this case is specifically for nodes that are deleted by an external source.

Our SIG has been having some discussions around "cloud provider E2E suite" which is likely going to help with this but it's not something we have setup yet (cc @d-nishi). Either way, I'll start exploring some options for this particular test case. I personally would prefer we merge this in for 1.13 since it's a pretty big PR and I don't want to deal with conflicts later, but I'll defer to your judgement. I think it's unlikely we'll have E2E coverage for this in time for v1.13 though.

**Comment by mtaufen** [comment]
Would the cluster autoscaler tests catch the deletion case?

**Comment by mtaufen** [comment]
Specifically this one: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/autoscaling/cluster_size_autoscaling.go#L656

**Comment by mtaufen** [comment]
@mwielgus @DirectXMan12 do you know if those tests can help cover this PR?

**Comment by mtaufen** [comment]
After thinking through it, I think we can merge this. The cases I'd be most worried about are unit tested for this controller, and the code change in the existing NLC was fairly isolated. 

Still curious about the autoscaler tests.

/lgtm

**Comment by andrewsykim** [comment]
Thanks @mtaufen! I'll dig deeper into the cluster autoscaler tests and follow up with you. 

@deads2k @lavalamp can we have your review here please? TL;DR we moved the "cloud" part of node lifecycle controller into its own controller so we can disable it without disabling the rest of the node lifecycle logic.

**Comment by nikopen** [comment]
We've hit code freeze since Friday and this is quite a big PR, I leave it up to @deads2k @lavalamp to decide the milestone

**Comment by andrewsykim** [comment]
@mtaufen had to make some minor tweaks, mind reviewing & re-applying lgtm?

**Comment by mtaufen** [comment]
@andrewsykim can you flag what the tweaks were?

File: pkg/controller/cloud/node_lifecycle_controller.go
Comment (cheftako) [inline pkg/controller/cloud/node_lifecycle_controller.go]: Would CloudNodeCleanupController be a better name?

Comment (mtaufen) [reply pkg/controller/cloud/node_lifecycle_controller.go]: It's also responsible for the shutdown taint, which doesn't really fit with all of the cleanup/garbage collection synonyms. We settled on this name in prior comments since it fits the current naming and makes it clear this is the cloud-specific part. Could even be we eventually just merge this in with the CCM stuff down the road.

**Comment by mtaufen** [comment]
/assign @deads2k 
for approval

**Review by mtaufen — APPROVED** [review]

**Comment by mtaufen** [comment]
/unassign @deads2k 
/assign @dchen1107 
for approval

**Comment by mtaufen** [comment]
@andrewsykim FYI I noticed some rebase artifacts (e.g. `>>>>>>>`) in your commits; do you mind just squashing them together?

**Comment by andrewsykim** [comment]
Sure thing, didn't initially so it's easier to review.

**Comment by andrewsykim** [comment]
friendly ping @deads2k @dchen1107 for review :)

**Pull request merged**
